include/hip/hcc_detail/ocml_math_fwd.h

```diff
@@ -0,0 +1,431 @@
+__attribute__((pure))
+float __ocml_pow_f32(float, float);
+__attribute__((pure))
+float __ocml_rcbrt_f32(float);
+__attribute__((const))
+float __ocml_remainder_f32(float, float);
+float __ocml_remquo_f32(float, float, __attribute__((address_space(5))) int*);
+__attribute__((const))
+float __ocml_rhypot_f32(float, float);
+__attribute__((const))
+float __ocml_rint_f32(float);
+__attribute__((const))
+float __ocml_rlen3_f32(float, float, float);
+__attribute__((const))
+float __ocml_rlen4_f32(float, float, float, float);
+__attribute__((const))
+float __ocml_round_f32(float);
+__attribute__((pure))
+float __ocml_rsqrt_f32(float);
+__attribute__((const))
+float __ocml_scalb_f32(float, float);
+__attribute__((const))
+float __ocml_scalbn_f32(float, int);
+__attribute__((const))
+int __ocml_signbit_f32(float);
+float __ocml_sincos_f32(float, __attribute__((address_space(5))) float*);
+float __ocml_sincospi_f32(float, __attribute__((address_space(5))) float*);
+float __ocml_sin_f32(float);
+__attribute__((pure))
+float __ocml_sinh_f32(float);
+float __ocml_sinpi_f32(float);
+__attribute__((const))
+float __ocml_sqrt_f32(float);
+float __ocml_tan_f32(float);
+__attribute__((pure))
+float __ocml_tanh_f32(float);
+float __ocml_tgamma_f32(float);
+__attribute__((const))
+float __ocml_trunc_f32(float);
+float __ocml_y0_f32(float);
+float __ocml_y1_f32(float);
+
+// BEGIN INTRINSICS
+__attribute__((const))
+float __ocml_add_rte_f32(float, float);
+__attribute__((const))
+float __ocml_add_rtn_f32(float, float);
+__attribute__((const))
+float __ocml_add_rtp_f32(float, float);
+__attribute__((const))
+float __ocml_add_rtz_f32(float, float);
+__attribute__((const))
+float __ocml_sub_rte_f32(float, float);
+__attribute__((const))
+float __ocml_sub_rtn_f32(float, float);
+__attribute__((const))
+float __ocml_sub_rtp_f32(float, float);
+__attribute__((const))
+float __ocml_sub_rtz_f32(float, float);
+__attribute__((const))
+float __ocml_mul_rte_f32(float, float);
+__attribute__((const))
+float __ocml_mul_rtn_f32(float, float);
+__attribute__((const))
+float __ocml_mul_rtp_f32(float, float);
+__attribute__((const))
+float __ocml_mul_rtz_f32(float, float);
+__attribute__((const))
+float __ocml_div_rte_f32(float, float);
+__attribute__((const))
+float __ocml_div_rtn_f32(float, float);
+__attribute__((const))
+float __ocml_div_rtp_f32(float, float);
+__attribute__((const))
+float __ocml_div_rtz_f32(float, float);
+__attribute__((const))
+float __ocml_sqrt_rte_f32(float);
+__attribute__((const))
+float __ocml_sqrt_rtn_f32(float);
+__attribute__((const))
+float __ocml_sqrt_rtp_f32(float);
+__attribute__((const))
+float __ocml_sqrt_rtz_f32(float);
+__attribute__((const))
+float __ocml_fma_rte_f32(float, float, float);
+__attribute__((const))
+float __ocml_fma_rtn_f32(float, float, float);
+__attribute__((const))
+float __ocml_fma_rtp_f32(float, float, float);
+__attribute__((const))
+float __ocml_fma_rtz_f32(float, float, float);
+
+__attribute__((const))
```

Comment:
Why are these 4 here?  I don't think they belong.

Reply:
They are there because e.g.: https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__DOUBLE.html. Do you have an actual technical concern here or is it just a thing?

Reply:
Sorry, I meant the 4 following __llvm_amdgcn_<fn>_f32 (and similarly for the two f64 below.  These should never be called directly from HIP.

Reply:
I don't believe that is true. Note that CUDA actually exposes intrinsics, amongst which there is e.g. https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__DOUBLE.html#group__CUDA__MATH__INTRINSIC__DOUBLE_1g76b42ce1d523e8655454a061d0ed3728, which are understood to be hardware specific. There's no reason for us not to use the equivalent capability; furthermore OCML doesn't expose things like reciprocal, and penalising the user by doing this as 1 / x appears counterproductive.

Reply:
HIP does not define and document a "native" API, so I'm not sure this is a good idea.  At least change the function names to something meaningful like "amd_native_xyz" instead of those deliberately ugly  __llvm_amdgcn_xyz_fMN.

Reply:
I'm not sure I follow. The user calls the HIP intrinsics, which MUST match the CUDA name. This is not a matter of whether or not it is a good idea, it is practically infeasible to not support them as ported code will break. As for the intrinsic, that's a forward declare that matches the actual name, just like all other forward declares in that header, and is an implementation detail - the user calls `__frcp_rd(x)` and not `__llvm_amdgcn_rcp_f32(x)`. Making something that is intended to signal "danger danger" cute would subvert the initial intent, ignoring the fact that it simply does not work - at some point something must call `__llvm_amdgcn_rcp_f32`, and a declaration of that must be in scope.

Reply:
I think this finally starts to answer my original question "Why are these 4 here".  And the answer appears to be "they are used in some other functions such as __frcp which we have to implement and are supposed to provide native ISA semantics".  Now I understand why these exist at all, but these still don't fit in a file named ocml_math_fwd.h.

Reply:
Brilliant! Yes,  the name of the header is suboptimal in that it is not generic. I've adjusted this, thanks.